File: CSharp/leetcodesln/ValidPalindrome.cs
namespace leetcodesln
{
    public class ValidPalindrome
    {
        public bool IsPalindrome(string s)
        {
            int l = 0, r = s.Length - 1;

            while (l < r)
            {
                while (l < r && !char.IsLetterOrDigit(s[l])) ++l;
                while (r > l && !char.IsLetterOrDigit(s[r])) --r;
                if (char.ToLower(s[l++]) != char.ToLower(s[r--])) return false;
            }
            return true;
        }
    }
}

File: Go/questions/ValidPalindrome.go
package questions

import "unicode"

func IsPalindrome(s string) bool {
	left, right := 0, len(s) -1
	r := []rune(s)
	for left < right {
		for  left < right && !isLetterOrDigit(unicode.ToLower(r[left]))  {
			left++
		}
		for right > left && !isLetterOrDigit(unicode.ToLower(r[right]))  {
			right--
		}
		if unicode.ToLower(r[left]) != unicode.ToLower(r[right]) {
			return false
		}
		left++
		right--
	}
	return true
}

func isLetterOrDigit(r rune) bool {
	return ('a' <= r && r <= 'z') || ('0' <=r && r <='9')
}
